File: matlab/+copilot/DefineTool.m
classdef DefineTool
    % DefineTool  Declare a tool that the Copilot agent can invoke.
    %
    %   tool = copilot.DefineTool( ...
    %       'Name',        'get_weather', ...
    %       'Description', 'Return weather for a city', ...
    %       'Parameters',  struct('type','object', ...
    %                             'properties', struct('city', struct('type','string')), ...
    %                             'required', {{'city'}}), ...
    %       'Handler',     @(params) getWeather(params.city));

    properties
        % Unique tool name (snake_case recommended).
        Name (1,:) char = ''

        % Human-readable description shown to the agent.
        Description (1,:) char = ''

        % JSON Schema describing the tool parameters (struct).
        Parameters (1,1) struct = struct('type', 'object', 'properties', struct())

        % Function handle invoked when the agent calls the tool.
        % Signature: result = handler(paramsStruct)
        Handler function_handle = @(~) struct()

        % When true, this tool overrides a built-in tool with the same name.
        OverridesBuiltInTool (1,1) logical = false

        % When true, the SDK will not request permission before execution.
        SkipPermission (1,1) logical = false
    end

    methods
        function obj = DefineTool(varargin)
            %DefineTool  Create a tool definition using name-value pairs.

            p = inputParser;
            p.addParameter('Name',                 obj.Name);
            p.addParameter('Description',          obj.Description);
            p.addParameter('Parameters',           obj.Parameters);
            p.addParameter('Handler',              obj.Handler);
            p.addParameter('OverridesBuiltInTool', obj.OverridesBuiltInTool);
            p.addParameter('SkipPermission',       obj.SkipPermission);
            p.parse(varargin{:});

            obj.Name                = p.Results.Name;
            obj.Description         = p.Results.Description;
            obj.Parameters          = p.Results.Parameters;
            obj.Handler             = p.Results.Handler;
            obj.OverridesBuiltInTool = p.Results.OverridesBuiltInTool;
            obj.SkipPermission      = p.Results.SkipPermission;
        end

        function s = toStruct(obj)
            %toStruct  Convert the tool to a struct suitable for JSON-RPC.
            s.name        = obj.Name;
            s.description = obj.Description;
            s.parameters  = obj.Parameters;
            if obj.OverridesBuiltInTool
                s.overridesBuiltInTool = true;
            end
            if obj.SkipPermission
                s.skipPermission = true;
            end
        end
    end
end

File: visualbasic/src/DefineTool.vb
' Copyright (c) Microsoft Corporation. All rights reserved.

Imports System.Text.Json

Namespace GitHub.Copilot.SDK

    ''' <summary>
    ''' Defines a custom tool that can be registered with a session so the
    ''' assistant can invoke it during a conversation.
    ''' </summary>
    ''' <example>
    ''' <code>
    ''' Dim weatherTool As New DefineTool With {
    '''     .Name = "get_weather",
    '''     .Description = "Get current weather for a city",
    '''     .ParametersSchema = New Dictionary(Of String, Object) From {
    '''         {"type", "object"},
    '''         {"properties", New Dictionary(Of String, Object) From {
    '''             {"city", New Dictionary(Of String, Object) From {
    '''                 {"type", "string"},
    '''                 {"description", "City name"}
    '''             }}
    '''         }},
    '''         {"required", New String() {"city"}}
    '''     },
    '''     .Handler = Async Function(input)
    '''                    Dim el = DirectCast(input, JsonElement)
    '''                    Dim city = el.GetProperty("city").GetString()
    '''                    Return $"Weather in {city}: 72F, sunny"
    '''                End Function
    ''' }
    ''' </code>
    ''' </example>
    Public Class DefineTool

        ''' <summary>Unique name of the tool (e.g. "get_weather").</summary>
        Public Property Name As String

        ''' <summary>Human-readable description shown to the model.</summary>
        Public Property Description As String

        ''' <summary>
        ''' JSON Schema describing the tool's input parameters.
        ''' Typically a Dictionary(Of String, Object) representing a JSON Schema object.
        ''' </summary>
        Public Property ParametersSchema As Object

        ''' <summary>
        ''' Async handler invoked when the assistant calls this tool.
        ''' The input parameter is the deserialized arguments (usually a <see cref="JsonElement"/>).
        ''' Return a string result for the model.
        ''' </summary>
        Public Property Handler As Func(Of Object, Task(Of String))

        ''' <summary>
        ''' Creates an empty tool definition. Set properties before registering.
        ''' </summary>
        Public Sub New()
        End Sub

        ''' <summary>
        ''' Creates a fully specified tool definition.
        ''' </summary>
        Public Sub New(name As String, description As String, parametersSchema As Object, handler As Func(Of Object, Task(Of String)))
            Me.Name = name
            Me.Description = description
            Me.ParametersSchema = parametersSchema
            Me.Handler = handler
        End Sub

    End Class

End Namespace
